### ISSUE TITLE
DRAM and AXI interconnect throughput with single-beat master

### ISSUE THREAD

**Issue by Juan-Gg** [openhwgroup/cva6#2036]
**TL; DR;**
I need axi single-beat transactions with good throughput. The axi_riscv_atomics module seems to limit the number of outstanding transactions.

**Context** 
I am integrating an AXI master into the CVA6 SoC. This master is required to access memory with a variable stride, so it is a single-beat design. It has support for multiple outstanding transactions. I am using the [axi_lite_to_axi](https://github.com/pulp-platform/axi/blob/master/src/axi_lite_to_axi.sv) adapter to connect to the crossbar.

I have tested it in the CVA6 SoC and observed the bus on my master side of the crossbar with an ILA, see attached. I can only get a new data word every 20 or so cycles, even when there are 5 outstanding (in-flight)  transactions. The behavior is the same when writing.

The AXI bus as seen on my accelerator's master  port
![image](https://github.com/openhwgroup/cva6/assets/29777196/492c73a3-9797-49b9-83d8-4a5f23397cf9)

**Problem**
I need more throughput than this for my target application.

**What I have done**
In the crossbar configuration, I have increased MaxMstTrans and MaxSlvTrans from 1 to 4, to allow for more in-flight/outstanding transactions. Even increasing it to 10 does not get me more than 5 outstanding transactions.

Below is a block diagram of everything between the crossbar and the DRAM. I have instrumented the valid and ready signals in between. It seems that it is the riscv_atomics module the one that's limiting throughput, not allowing new read requests until the previous request completes.

This module has a parameter, "AXI_MAX_WRITE_TXNS" ( Maximum number of AXI write bursts outstanding at the same time ), but that would only affect write transactions, not reads? I have increased it with no change in throughput.

Modules between crossbar and DRAM:
<img src="https://github.com/openhwgroup/cva6/assets/29777196/d973dc6c-5408-4a17-ad73-ce37d79417a9" width=60% height=60%>

AXI valid and ready signals between crossbar, riscv_atomics, clock_converter and memory controller, in the same order as above block diagram:
![image](https://github.com/openhwgroup/cva6/assets/29777196/950044b7-ad3c-4516-a20a-2f716d63407c)
(Note that s_axi_... is in another clock domain, so there is not much to be trusted from those traces)

**Any ideas ?**

Given than there is a "// Probably requires update" comment next to the MaxMstTrans and MaxSlvTrans, set to 1 on the crossbar configuration, I assume the number of outstanding transactiosn has not been given much thought.

Any ideas are welcome.

Thank you,
           Juan

**Additional information:**

A related [article from ZipCPU](https://zipcpu.com/blog/2020/03/23/wbm2axisp.html) talks about throughput of the xilinx memory controller:
_The good news for all of these applications is that [Xilinx’s SDRAM MIG controller](https://www.xilinx.com/support/documentation/ip_documentation/mig_7series/v1_4/ug586_7Series_MIS.pdf) controller can handle single beat random access requests for reads or writes (not both) with 100% throughput–depending of course upon your access pattern._

**Comment by Juan-Gg** [comment]
Indeed, it was the riscv_atomics module. It turns out, as with other submodules on this project, it was a 2019 version and significant work has been carried out since.

I updated the corev_apu/src/axi_riscv_atomics [submodule](https://github.com/pulp-platform/axi_riscv_atomics) with the most recent version (v0.8.2 at the time of writing), just running `git checkout master`. Support for multiple read and write outstanding transactions was added in v0.3.0.

I just had to configure the new AXI_MAX_READ_TXNS, which together with AXI_MAX_WRITE_TXNS set the maximum number of read and write outstanding transactions that this module allows. The additional parameters of this newer version can be left with their default values.

For compatibility with the current axi submodule version:
- axi_riscv_atomics.sv: axi_multicut instance: change axi_req_t for req_t and axi_resp_t for resp_t. 

Here are the waveforms now, with the maximum 6 outstanding transactions that I have set in my own accelerator being the limit at the moment. With about 24 cycles latency, on average, each transaction takes 4 to 5 cycles.
![image](https://github.com/openhwgroup/cva6/assets/29777196/5a70099c-b4b2-4fa6-a24a-118bbd1ffc46)

As a summary of things influencing the number of outstanding transactions:
- The axi master's capabilities.
- AXI crossbar (axi_xbar_intf): Configurable with the MaxMstTrans and MaxSlvTrans parameters.
- RISCV atomics (axi_riscv_atomics_wrap): Configurable with the AXI_MAX_READ_TXNS and AXI_MAX_WRITE_TXNS parameters, after updating its version.
- The memory controller. Not configurable that I'm aware of.
